### PR TITLE
Adding Intel nGraph build

### PR DESCRIPTION
**Pull request by baojun-nervana**

### CODE REVIEW

**Comment by baojun-nervana** [comment]
@luotao1 Cannot assign reviewer. This is the ngraph cmake PR

**Comment by luotao1** [comment]
please add `test=develop` in commit information to trigger the CI

File: Dockerfile [hunk 1]
```diff
@@ -11,11 +11,13 @@ RUN /bin/bash -c 'if [[ -n ${UBUNTU_MIRROR} ]]; then sed -i 's#http://archive.ub
 # ENV variables
 ARG WITH_GPU
 ARG WITH_AVX
+ARG WITH_NGRAPH
 ARG WITH_DOC
 
 ENV WOBOQ OFF
 ENV WITH_GPU=${WITH_GPU:-ON}
 ENV WITH_AVX=${WITH_AVX:-ON}
+ENV WITH_NGRAPH=${WITH_NGRAPH:-OFF}
```
Comment (luotao1) [inline Dockerfile]: 这个文件可以不用更改

File: cmake/external/ngraph.cmake [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    CMAKE_ARGS          -DMKLDNN_LIB_DIR=${MKLDNN_INSTALL_DIR}/lib
+)
+
+# Workaround for nGraph expecting mklml to be in mkldnn install directory.
+ExternalProject_Add_Step(
+    ${NGRAPH_PROJECT}
+    PrepareMKL
+    COMMAND ${CMAKE_COMMAND} -E copy_directory ${MKLML_INSTALL_DIR}/lib ${MKLDNN_INSTALL_DIR}/lib
+    DEPENDEES download
+    DEPENDERS configure
```
Comment (luotao1) [inline cmake/external/ngraph.cmake:85]: 75行，这里可以拷贝一个软连接么？或者ngraph后面能支持单独的mklml目录么？

File: paddle/scripts/paddle_build.sh [hunk 3]
```diff
@@ -500,7 +502,8 @@ EOF
         -DCMAKE_BUILD_TYPE=Release \
         -DWITH_DOC=ON \
         -DWITH_GPU=OFF \
-        -DWITH_MKL=OFF
+        -DWITH_MKL=OFF \
+        -DWITH_NGRAPH=OFF
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh]: 506行不用加，因为默认是OFF

File: paddle/scripts/paddle_build.sh [hunk 4]
```diff
@@ -684,7 +688,7 @@ function test_fluid_lib() {
     ========================================
 EOF
         cd ${PADDLE_ROOT}/paddle/fluid/inference/api/demo_ci
-        ./run.sh ${PADDLE_ROOT} ${WITH_MKL:-ON} ${WITH_GPU:-OFF} ${INFERENCE_DEMO_INSTALL_DIR} \
+        ./run.sh ${PADDLE_ROOT} ${WITH_MKL:-ON} ${WITH_NGRAPH:-OFF} ${WITH_GPU:-OFF} ${INFERENCE_DEMO_INSTALL_DIR} \
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh]: 691行不用加

File: paddle/scripts/paddle_build.sh [hunk 5]
```diff
@@ -171,6 +172,7 @@ EOF
         -DWITH_AMD_GPU=${WITH_AMD_GPU:-OFF} \
         -DWITH_DISTRIBUTE=${WITH_DISTRIBUTE:-OFF} \
         -DWITH_MKL=${WITH_MKL:-ON} \
+        -DWITH_NGRAPH=${WITH_NGRAPH:-OFF} \
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:175]: 142/175行先把OFF打开，让ci跑NGRAPH=ON的试下，如果通过：
- 在PR里再改成OFF
- CI配置里，我们会配置成ON

File: Dockerfile (hunk 1 above)
Comment (baojun-nervana) [reply Dockerfile]: will remove

Comment (baojun-nervana) [reply Dockerfile]: will remove

File: cmake/external/ngraph.cmake (hunk 2 above)
Comment (baojun-nervana) [reply cmake/external/ngraph.cmake:85]: Let us looking into this. Currently we use whatever DO uses.

Comment (baojun-nervana) [reply cmake/external/ngraph.cmake:85]: Used symlink

**Comment by luotao1** [comment]
1. please add `test=develop` in commit information to trigger the CI
2. please not use `gith push -f`

**Comment by baojun-nervana** [comment]
@luotao1 Can you help to look at the python35 fail?

**Comment by luotao1** [comment]
python35本地自测方法：
- 下载镜像
```
docker pull paddlepaddle/paddle_manylinux_devel:cuda8.0_cudnn7 
```
- 设置环境变量
```
export LD_LIBRARY_PATH=/opt/_internal/cpython-3.5.1/lib/:${LD_LIBRARY_PATH}
export PATH=/opt/_internal/cpython-3.5.1/bin/:${PATH}
export PYTHON_FLAGS="-DPYTHON_EXECUTABLE:FILEPATH=/opt/_internal/cpython3.5.1/bin/python3 -DPYTHON_INCLUDE_DIR:PATH=/opt/_internal/cpython-3.5.1/include/python3.5m -DPYTHON_LIBRARIES:FILEPATH=/opt/_internal/cpython-3.5.1/lib/libpython3.so"
export PYTHON_ABI="cp35-cp35m"
export WITH_FLUID_ONLY=ON
export PY_VERSION=3.5
```
- 编译
```
bash paddle/scripts/paddle_build.sh build
```

**Comment by luotao1** [comment]
@shanyi15 @velconia 虽然有[python35安装](http://www.paddlepaddle.org/documentation/docs/zh/1.1/beginners_guide/install/compile/compile_CentOS.html)，能否上一个文档，说明如何复现CI环境错误呢？

**Comment by velconia** [comment]
这个错误出自于 ```ngraph``` 这个external project编译出错了，复现CI环境的方法可以参考这篇[文档](http://www.paddlepaddle.org/documentation/docs/zh/1.1/beginners_guide/install/compile/compile_CentOS.html)

File: paddle/scripts/paddle_build.sh [hunk 6]
```diff
@@ -171,6 +172,7 @@ EOF
         -DWITH_AMD_GPU=${WITH_AMD_GPU:-OFF} \
         -DWITH_DISTRIBUTE=${WITH_DISTRIBUTE:-OFF} \
         -DWITH_MKL=${WITH_MKL:-ON} \
+        -DWITH_NGRAPH=${WITH_NGRAPH:-ON} \
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh]: 142和175行可以改成OFF，PR merge后，会在CI上开成ON的。

File: paddle/scripts/paddle_build.sh [hunk 7]
```diff
@@ -525,6 +527,7 @@ EOF
         -DWITH_DOC=ON \
         -DWITH_GPU=OFF \
         -DWITH_MKL=OFF \
+        -DWITH_NGRAPH=OFF \
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh]: 530行可以去掉

**Review by luotao1 — COMMENTED** [review]
LGTM

**Comment by baojun-nervana** [comment]
see to be random fail.
PY2:
[20:50:15] :	 [Step 1/1] 99% tests passed, 1 tests failed out of 468
[20:50:15] :	 [Step 1/1] 
[20:50:15] :	 [Step 1/1] Total Test time (real) = 2128.08 sec
[20:50:15] :	 [Step 1/1] 
[20:50:15] :	 [Step 1/1] The following tests FAILED:
[20:50:15] :	 [Step 1/1] 	212 - test_gru_op (Timeout)
PY3:
[21:02:32] :	 [Step 1/1] 99% tests passed, 1 tests failed out of 448
[21:02:32] :	 [Step 1/1] 
[21:02:32] :	 [Step 1/1] Total Test time (real) = 1800.40 sec
[21:02:32] :	 [Step 1/1] 
[21:02:32] :	 [Step 1/1] The following tests FAILED:
[21:02:32] :	 [Step 1/1] 	376 - test_py_reader_using_executor (OTHER_FAULT)

**Review by luotao1 — APPROVED** [review]
LGTM

File: cmake/external/ngraph.cmake [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+)
+
+if(UNIX AND NOT APPLE)
+    include(GNUInstallDirs)
+    SET(NGRAPH_LIB_DIR ${NGRAPH_INSTALL_DIR}/${CMAKE_INSTALL_LIBDIR})
+else()
+    SET(NGRAPH_LIB_DIR ${NGRAPH_INSTALL_DIR}/lib)
+endif()
+MESSAGE(STATUS "nGraph lib will be installed at: ${NGRAPH_LIB_DIR}")
+
```
Comment (tensor-tang) [inline cmake/external/ngraph.cmake:73]: what do you mean by this "if ... else ..." here?

File: python/setup.py.in [hunk 9]
```diff
@@ -174,6 +174,18 @@ if '${CMAKE_BUILD_TYPE}' == 'Release':
             raise Exception("patch libmkldnn.so failed, command: %s" % command)
         package_data['paddle.libs']+=['libmkldnn.so.0']
         shutil.copy('${MKLDNN_SHARED_LIB}', libs_path)
+if '${WITH_NGRAPH}' == 'ON':
+    if '${CMAKE_BUILD_TYPE}' == 'Release':
```
Comment (tensor-tang) [inline python/setup.py.in:178]: This is duplicated with  Line163, could be placed in the Line 163

File: cmake/external/ngraph.cmake (hunk 8 above)
Comment (baojun-nervana) [reply cmake/external/ngraph.cmake:73]: addressed in PR https://github.com/PaddlePaddle/Paddle/pull/14441

File: python/setup.py.in (hunk 9 above)
Comment (baojun-nervana) [reply python/setup.py.in:178]: addressed in PR https://github.com/PaddlePaddle/Paddle/pull/14441

**Review by baojun-nervana — COMMENTED** [review]
review addressed https://github.com/PaddlePaddle/Paddle/pull/14441

**Pull request merged**